pkg/services/blockfetcher/blockfetcher.go

```diff
@@ -0,0 +1,353 @@
+			prm.SetFilters(filters)
+
+			blockOids, err := bfs.search(prm)
+			bfs.log.Info(fmt.Sprintf("Found %d blocks from index %d to %d", len(blockOids), startIndex, startIndex+batchSize-1))
+			if err != nil {
+				bfs.log.Error(fmt.Sprintf("Failed to find block with index %d", startIndex), zap.Error(err))
+				return err
+			}
+
+			if len(blockOids) == 0 {
+				bfs.log.Info(fmt.Sprintf("No block found with index %d, stopping.", startIndex))
+				return errors.New("no block found")
+			}
+			for _, oidBlock := range blockOids {
+				go func(oidBlock oid.ID) {
+					data, err := bfs.get(oidBlock.String())
+					if err != nil {
+						bfs.log.Error(fmt.Sprintf("Failed to fetch block %s", oidBlock.String()), zap.Error(err))
+					}
+					err = bfs.processBlock(data)
+					if err != nil {
+						bfs.log.Error(fmt.Sprintf("Failed to process block with index %d", startIndex), zap.Error(err))
+					}
+				}(oidBlock)
+			}
+
+			startIndex += batchSize
+		}
+	}
+}
+
+func (bfs *Service) fetchDataWithOid() error {
+	startIndex := bfs.chain.BlockHeight() / 2000
+	skip := bfs.chain.BlockHeight() % 2000
+
+	for {
+		select {
+		case <-bfs.quit:
+			return nil
+		default:
+			prm := client.PrmObjectSearch{}
+			filters := object.NewSearchFilters()
+			filters.AddFilter("block_oids", fmt.Sprintf("%d", startIndex), object.MatchNumGE)
+			filters.AddFilter("block_oids", fmt.Sprintf("%d", startIndex), object.MatchNumLE)
+			prm.SetFilters(filters)
+
+			blockOidsObject, err := bfs.search(prm)
+			if err != nil {
+				bfs.log.Error(fmt.Sprintf("Failed to find 'block_oids' object with index %d", startIndex), zap.Error(err))
+				return err
+			}
+			if len(blockOidsObject) == 0 {
+				bfs.log.Info(fmt.Sprintf("No 'block_oids' object found with index %d, stopping.", startIndex))
+				return errors.New("no 'block_oids' object found")
+			}
+
+			blockOidsData, err := bfs.get(blockOidsObject[0].String())
+			if err != nil {
+				bfs.log.Error(fmt.Sprintf("Failed to fetch 'block_oids' object %d", startIndex), zap.Error(err))
+				return err
+			}
+			blockOIDs, err := parseBlockOIDs(blockOidsData)
+			if err != nil {
+				bfs.log.Error(fmt.Sprintf("Failed to parse 'block_oids' object with index %d", startIndex), zap.Error(err))
+				return err
+			}
+
+			for j, oidBlock := range blockOIDs {
+				if j < int(skip) {
+					continue
+				}
+				data, err := bfs.get(oidBlock.String())
+				if err != nil {
+					bfs.log.Error(fmt.Sprintf("Failed to fetch block with OID %s", oidBlock.String()), zap.Error(err))
+					return err
+				}
+				bfs.log.Info(fmt.Sprintf("Fetched block with OID %s", oidBlock.String()))
+				err = bfs.processBlock(data)
+				if err != nil {
+					bfs.log.Error(fmt.Sprintf("Failed to process blocks with index %d", startIndex), zap.Error(err))
+					return err
+				}
+			}
+
+			startIndex++
+		}
+	}
+}
+
+func (bfs *Service) processBlock(data []byte) error {
+	var buf []byte
+	stateRootInHeader := bfs.chain.GetConfig().StateRootInHeader
```

Comment:
Fetch this value once in service constructor and save it into service's config. Don't make it every time you save a block.